### ISSUE TITLE
Query: missing variable #4 (my_variable_name) for node CalculationNode while planning registers

### ISSUE THREAD

**Issue by egilewski** [arangodb/arangodb#2415]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [x] 3.1
- [ ] 3.2 pre-releases: ___
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
- [x] I'm using graph features

I'm issuing AQL via:
- [x] web interface with this browser: Firefox ESR 45.8.0 running on this OS: Debian Testing
- [ ] arangosh
- [ ] this Driver:

I've run `db._explain("<my aql query>")` and it didn't shed more light on this (error is basically the same).
The AQL query in question is:

```
LET requested_user = 'customers/0'
LET products_requested_user_viewed = (
    FOR product IN OUTBOUND requested_user views
    RETURN product
)
LET users_which_viewed_same_products = (
    FOR product IN products_requested_user_viewed
        FOR user in INBOUND product views
            FILTER user._id != requested_user
            RETURN DISTINCT user
)
FOR user in users_which_viewed_same_products
    FOR product in OUTBOUND user views
        FILTER product NOT IN products_requested_user_viewed
        COLLECT p = product INTO views_by_product
        SORT LENGTH(views_by_product) DESC
        RETURN {id: p._id, views_count: LENGTH(views_by_product)}
```

It fails with `Query: missing variable #4 (products_requested_user_viewed) for node CalculationNode while planning registers (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/Aql/ExecutionNode.cpp:1133). Please report this error to arangodb.com (while executing) (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/RestHandler/RestCursorHandler.cpp:131). Please report this error to arangodb.com`

However next query works (the only change is that the last query (not sure about the right term) became a `LET`, and the last two lines were added):

```
LET requested_user = 'customers/0'
LET products_requested_user_viewed = (
    FOR product IN OUTBOUND requested_user views
    RETURN product
)
LET users_which_viewed_same_products = (
    FOR product IN products_requested_user_viewed
        FOR user in INBOUND product views
            FILTER user._id != requested_user
            RETURN DISTINCT user
)
LET products_those_users_viewed = (
    FOR user in users_which_viewed_same_products
        FOR product in OUTBOUND user views
            // Should be optional or just lower the weight.
            FILTER product NOT IN products_requested_user_viewed 
            COLLECT p = product INTO views_by_product
            SORT LENGTH(views_by_product) DESC
            RETURN {id: p._id, views_count: LENGTH(views_by_product)}
)
FOR product IN products_those_users_viewed
    RETURN product
```

The issue can be reproduced using this dataset:

https://gist.github.com/egilewski/08c702ec4e27b193ed2cd2dfcb1b9d54

P.S. This is my first graph query, so I'm not even ashamed if it's awful.

**Comment by jsteemann** [comment]
Obviously this error should not happen.
I debugged it and found that it was due to subquery inlining. After a subquery was inlined, the required reassessment of the execution plan was not triggered immediately, but only after the inlining optimizing rule. However, for this particular query, the subquery inlining optimizer rule was about to inline a second subquery, and this required the reassessment of the execution plan. I have added these steps in commit 7a9cae1cb00b5c061b421f2bc3651ce94829b3d4.
So the query should work without any issues now. The fix has been pushed into the 3.1 and devel branches. It will become available with the next 3.1 release.
